Update second passphrase UI to align with sign in and registration - Closes #2537

colado opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### What issue have I solved?
<!--- Complementary description if needed -->
#2537

### How have I implemented/fixed it?
- Refactored `passphraseRenderer` to accept passphrase confirmation logic.
- Applied passphrase confirmation logic to 2nd passphrase.
- The last step of 2nd passphrase registration does not contain the box since this component will be changed in a separate issue.

### How has this been tested?
Made sure passphrase creation is successful and screens aligned with design


### Review checklist
- The PR follows our [Test guide](/LiskHQ/lisk-hub/blob/development/docs/TEST_GUIDE.md)
- The PR follows our [CSS guide](/LiskHQ/lisk-hub/blob/development/docs/CSS_GUIDE.md)

slaweet commented on src/components/screens/secondPassphrase/confirmPassphrase.js:
```diff
@@ -0,0 +1,28 @@
+      <div className={styles.passphraseConfirmationContainer}>
+        <p className={styles.info}>{t('Based on your passphrase that was generated in the previous step, select the missing words below')}</p>
+        <PassphraseRenderer
+          values={passphrase.split(' ')}
```
```suggestion
          passphrase={passphrase}
```
💅 It would be DRYer to pass `passphrase` and call `.split(' ')` inside `PassphraseRenderer` than here and  in https://github.com/LiskHQ/lisk-hub/pull/2575/files#diff-618c1dc31da4cb19645765d9e46e985eR52

slaweet commented on src/components/screens/secondPassphrase/secondPassphrase.css:
```diff
@@ -54,3 +62,27 @@
+
+.lastStep {
+  min-width: 720px;
+}
```
💅 This class can be replaced by https://github.com/LiskHQ/lisk-hub/blob/1a8d62a40fbdabeab04e063189d41cb536137bc9/src/components/toolbox/box/demo.js#L19

colado replied:
✅

slaweet commented on src/components/screens/secondPassphrase/secondPassphrase.test.js:
```diff
@@ -60,45 +66,54 @@ describe('SecondPassphrase', () => {
 
-    it('should handle registerSecondPassphrase failure', () => {
+    // TODO enable the next test when implemented
+    it.skip('should handle registerSecondPassphrase failure', () => {
```
🐛 Failure path not implemented

colado replied:
✅

slaweet commented on src/components/screens/secondPassphrase/summaryStep.js:
```diff
@@ -22,6 +21,8 @@ const SummaryStep = ({
           account: account.info.LSK,
           callback: ({ success, error }) => {
+            // TODO: To be tackled during multiStep refactoring
+            /* istanbul ignore next */
```
AFAIK this is just because of the skipped test case for failure path. Failure path should IMO be part of this PR, and then it should be fine to remove this comment.

colado replied:
I have implemented the failure path but I still need design to provide the final copy and to let me know if there is anything else there that would need to change in case of failure so I will get back to this when I hear back from them.

colado replied:
✅

slaweet commented on src/components/shared/passphraseBackup/index.test.js:
```diff
@@ -0,0 +1,20 @@
+  it('should show tip after copying passphrase', () => {
+    wrapper = mount(<PassphraseBackup {...props} />);
+    wrapper.find('CopyToClipboard').at(0).simulate('click');
+    expect('tip').toBeTruthy();
```
💅 This assertion will be always true, no matter what you'll do with the component. It's not checking what's in `wrapper`

colado replied:
I made the changes to check the text of the tip but somehow even when `this.state.showTip` is true, the test returns the content of the `tip` as empty. Can you check if there is something missing on the test?

slaweet replied:
This was the problem: https://github.com/LiskHQ/lisk-hub/pull/2575/commits/8f2431b286c478a392acd6961aa20ddf2c1eda35

I also did two more commits on this file. See commit messages for more info...

slaweet commented on src/components/shared/passphraseRenderer/index.js:
```diff
@@ -0,0 +1,231 @@
+    this.state = {
+      indexes: [2, 9],
+      fieldSelected: undefined,
+      displayOptions: undefined,
```
🐛 (or 📚?)  Based on the designs, I think the initial state should be the first missing word selected. There is no screen in the designs where both missing words are empty, none of them is selected, and no options are displayed. In that case `state.displayOptions` is not needed, because it should depend on `state.fieldSelected`

This also applies to the previous Story https://github.com/LiskHQ/lisk-hub/pull/2556

colado replied:
You're right. I made the changes. It should also be possible to reselect an option once you've already selected it so I changed that also.

slaweet commented on src/components/shared/passphraseRenderer/index.js:
```diff
@@ -0,0 +1,231 @@
+      options: {},
+      isCorrect: false,
+      hasErrors: false,
+      disabledButton: true,
```
💅 I think it's not necessary to store this value, it can be easily computed from other state values

colado replied:
✅

slaweet replied:
I see you stopped using `disabledButton`, but forgot to remove this line
```suggestion
```

colado replied:
✅

slaweet commented on src/components/shared/passphraseRenderer/index.js:
```diff
@@ -0,0 +1,231 @@
+    this.handleConfirm = this.handleConfirm.bind(this);
+  }
+
+  UNSAFE_componentWillMount() { // eslint-disable-line camelcase
```
💅 IMO logic in this method should be moved to `constructor` since `UNSAFE_componentWillMount` is deprecated.

colado replied:
You're right. Although from what I see there's no reason it shouldn't work inside `componentDidMount` instead of `componentWillMount`. I moved it there, what do you think?

slaweet replied:
I think it's less code and more clear to the reader if it's all in one space the `constructor`.

slaweet commented on src/components/shared/passphraseRenderer/index.js:
```diff
@@ -0,0 +1,231 @@
+    this.timeout = setTimeout(cb, 1500);
+  }
+
+  getRandomIndexesFromPassphrase(qty) {
```
💅 naming a method `getSomething` suggests it will `return` something, so I'd suggest:
```suggestion
  setRandomIndexesFromPassphrase(qty) {
```

colado replied:
✅

slaweet commented on src/components/shared/passphraseRenderer/index.js:
```diff
@@ -0,0 +1,231 @@
+    const { values } = this.props;
+
+    const answers = Object.values(chosenWords);
+    const status = answers.filter((answer, index) => answer === values[indexes[index]])
```
💅 IMO more descriptive name would be
```suggestion
    const isCorrect = answers.filter((answer, index) => answer === values[indexes[index]])
```

colado replied:
✅

slaweet commented on src/components/shared/passphraseRenderer/index.js:
```diff
@@ -0,0 +1,231 @@
+      answers: [],
+      hasErrors: false,
+      chosenWords: {},
+      fieldSelected: undefined,
```
Here same is in the comment above

colado replied:
Which comment?

slaweet replied:
https://github.com/LiskHQ/lisk-hub/pull/2575#discussion_r339495583

colado replied:
✅

slaweet commented on src/components/shared/passphraseRenderer/index.js:
```diff
@@ -0,0 +1,231 @@
+      this.setState({
+        fieldSelected: undefined,
+        displayOptions: undefined,
+        chosenWords: {},
```
I don't see why this is needed. AFAIK `hasErrors` is in `state`, not in `props`.

colado replied:
True. I forgot to remove it after moving the logic from the screen to the component.

colado commented:
Blocked until missing copy is received

slaweet commented on src/components/shared/passphraseRenderer/index.js:
```diff
@@ -0,0 +1,226 @@
+
+    const cb = isCorrect
+      ? () => this.props.nextStep()
+      : () => this.setRandomIndexesFromPassphrase(2);
```
One more thought about making this even DRYer. If you define `const qty = 2` inside `setRandomIndexesFromPassphrase`, then you don't need to create new functions here and use just:

```js
    const cb = isCorrect ? this.props.nextStep : this.setRandomIndexesFromPassphrase;
```

slaweet commented on src/components/shared/passphraseRenderer/index.js:
```diff
@@ -0,0 +1,226 @@
+  setRandomIndexesFromPassphrase(qty) {
+    const { values } = this.state;
+    let idxs = values.map((w, index) => index);
+    const indexes = [...Array(qty)]
```
the `qty` variable name is not very descriptive, could be something like:
```suggestion
    const indexes = [...Array(numberOfMissingWords)]
```

colado replied:
✅

slaweet commented on src/components/shared/passphraseRenderer/index.js:
```diff
@@ -0,0 +1,225 @@
+      options: {
+        [indexes[0]]: options[0],
+        [indexes[1]]: options[1],
+      },
```
These two look very similar (not DRY):
https://github.com/LiskHQ/lisk-hub/blob/a2f966a9867a3dd7985f41d47bbe8ffbb63a04a8/src/components/shared/passphraseRenderer/index.js#L74-L80
https://github.com/LiskHQ/lisk-hub/blob/a2f966a9867a3dd7985f41d47bbe8ffbb63a04a8/src/components/shared/passphraseRenderer/index.js#L28-L34

It could be resolved if `this.assembleWordOptions` already returned `options` in the format that is used in `state`

slaweet commented on src/components/shared/passphraseRenderer/index.js:
```diff
@@ -0,0 +1,206 @@
+      indexes: initialIndexes,
+      fieldSelected: initialIndexes[0],
+      chosenWords: {},
+      options: this.assembleWordOptions(props.passphrase.split(' '), initialIndexes),
```
Solved in https://github.com/LiskHQ/lisk-hub/pull/2575/commits/755574759515149c0c5cdbe22a31827a2f130591

colado commented:
Thank you, @Efefefef. 
- Paper wallet title changed.
- Title on summary step changed.

Regarding the final step box, I have spoken with Julian and it was agreed to leave this as it is, since they have redesigned the whole `TransactionResult` component (the one used in the last step), so it will be changed in a separate issue for all the transactions. I updated the description here for clarification.
cc @imdominic.

jooltr commented:
Sorry for the delay guys, I thought it was done.

🎨Please double check the height of the passphrase container, it seems a bit smaller than the one provided and it feels clustered.

🎨Please double check the spacing between the line and the numbers, it seems a bit longer than the one provided and it feels clustered.

Thanks

colado commented:
Thank you, @jooltr. I have added some vertical padding to the passphrase container and fixed the space between the numbers and lines. Let me know if it looks ok to you now.

jooltr commented:
@colado the 2nd passphrase registration looks great! Thank you.

@sridharmeganathan please note that the feedback posted above is still relevant for the  passphrase container in the `Create an Account` screen.

Thanks!!
